Enhancement: Render / Turntable

BigRoy opened the pull request:
### Feature

Implements a draft for playblasting/rendering using the USD api.

---

_Opening this draft so I can comment on some code as we go_

Sasbom committed:
first commit, adding playblast module

Sasbom committed:
added utility and expanded on renderplayblast args and processing

Sasbom committed:
parsing complexity argument better, fixed walrus bugs

Sasbom committed:
slight refactor

Sasbom committed:
ConverFramePlaceholderblabla does internal None checking, so removed extra checks.

Sasbom committed:
Merge branch 'BigRoy:main' into enhancement/usdview_render_turntable

Sasbom committed:
Removed hardcoded complexity presets, added sanity checking for render engine arg

Sasbom committed:
Complete framerecorder initialization setup

Sasbom committed:
Merge branch 'BigRoy:main' into enhancement/usdview_render_turntable

Sasbom committed:
resources_rc added

Sasbom committed:
added menu, functionality for snapshot. Doesn't work as intended within hython 19.5.

Sasbom committed:
snapshot fix! Works now. Still needs major cleanup.

Sasbom committed:
Merge branch 'BigRoy:main' into enhancement/usdview_render_turntable

Sasbom committed:
changed logic in case no camera is selected to match usdrecord

Sasbom committed:
Merge branch 'enhancement/usdview_render_turntable' of https://github.com/Sasbom/usd-qtpy into enhancement/usdview_render_turntable

BigRoy commented on usd_qtpy/render_util/playblast.py:
```diff
@@ -0,0 +1,166 @@
+# NOTES:
+# pxr.UsdViewq.ExportFreeCameraToStage will export the camera from the view (a FreeCamera/ pxr.Gf.Camera, purely OpenGL)
+
+from pxr import Usd, UsdGeom
+from pxr import UsdAppUtils
+from pxr import Tf, Sdf
+
+from pxr.Usdviewq.stageView import StageView
+from viewer import CustomStageView
+
+from qtpy import QtCore
+
+from typing import Union
+import logging
```
Can we reorder the imports to be:

- first standard built-ins
- dependencies
- local imports

Also ensure local relative imports are forced to be relative imports via e.g. `.viewer import CustomStageView`.

```suggestion
import logging
from typing import Union

from qtpy import QtCore
from pxr import Usd, UsdGeom
from pxr import UsdAppUtils
from pxr import Tf, Sdf
from pxr.Usdviewq.stageView import StageView

from .viewer import CustomStageView
```

Sasbom replied:
I reordered everything, it makes sense.

`from .viewer import CustomStageView` will actually not resolve as I have defined render_util as a complete module.
I plan to add turntable functionality in a seperate file later, so I wanted to keep those seperate.

```py
from ..viewer import CustomStageView
```
Does resolve, so I will leave it as such.

BigRoy commented on usd_qtpy/editor.py:
```diff
@@ -94,5 +102,23 @@ def update_panel_checkstate():
+        # Render menu
+
+        render_menu = menubar.addMenu("Render")
+        render_labels = "Playblast" , "Snapshot"
```
PEP8 - check your linter. :) 
```suggestion
        render_labels = "Playblast", "Snapshot"
```

Sasbom replied:
> PEP8 - check your linter. :)

I really should ; _ ;

On the topic of PEP8 on a different note, I know that snake_case is preferable for functions, 
I just look at Qt and Usd and can't help but copy their style to maintain consistency.
That's my reasoning for naming functions the way I do. I will rename them all to be snake case, I will never more be tempted by the demons hiding in libraries.

BigRoy commented on usd_qtpy/editor.py:
```diff
@@ -94,5 +102,23 @@ def update_panel_checkstate():
+        goal_file = R"C:\dump\playblastview##.##.png"
+        
+        def render_snapshot(stage,stageview,path):
+            cam = render_util.playblast.cameraFromView(stage,stageview)
```
I feel like if this is complex or requires a function to access the camera we should just expose it on the `stageview` instead:

```
cam = stageview.get_active_camera()
```

Sasbom replied:
This is a really weird function, from the stageview's end.

It basically puts a camera in the hierarchy of the scene so I am not at all sure what to do with all of that yet or where it should belong. It doesn't keep a reference to the current viewer camera, because the viewer camera isn't a USD cam, it's a "FreeCam" that basically wraps an openGL camera.

I have put it in the render_util because my rationale is, that maybe it'd be neat to have a seperate "camera from current view" button to use. I would need to set the internals of the camera to make sense for a default 16:9 camera though, as currently it will spawn a camera with a "sensor size" based on the widget size.

So, currently, I just don't quite know where to put it. All I do know is that I want the resulting camera from it,
so I can later hook it up to a turntable, for example.

Maybe this could benefit from a context wrapper, like
```py
with currentViewCam(stageView) as cam:
    do_stuff_wth_cam(cam)
```
and it'd delete the camera from the hierarchy later on.

I don't really know yet. We'll see.

BigRoy commented on usd_qtpy/editor.py:
```diff
@@ -94,5 +102,23 @@ def update_panel_checkstate():
+            cam = render_util.playblast.cameraFromView(stage,stageview)
+            render_util.playblast.renderPlayblast(stage,path,"1",1920,cam,renderer="GL")
+
+        render_snap = partial(render_snapshot,self._stage,self._stageview,goal_file)
```
PEP8 check your linter; these are minor things though.
```suggestion
        render_snap = partial(render_snapshot, self._stage, self._stageview, goal_file)
```

Sasbom commented:
Thank you for opening the draft!
Seems like a good idea.

BigRoy commented on usd_qtpy/render_util/playblast.py:
```diff
@@ -0,0 +1,166 @@
+            frameRecorder.Record(stage, camera, timeCode, currentframe)
+            print("Done!")
+        except Tf.ErrorException as e:
+            logging.error("Recording aborted due to the following failure at time code {0}: {1}".format(timeCode, str(e)))
```
```suggestion
            logging.error("Recording aborted due to the following failure at time code %s: %s", timeCode, e)
```

If you format with a logger use `%s` and add in the variables to be formatted as extra arguments. See the suggestion.

This makes it so that the string is only formatted IF the error is handled by a logger handler - if not, then it doesn't pay for the time spent formatting the variables in the string. Errors are more likely handled by default, but still it's good optimization practice.

If you _need_ to format ahead of time since you want to format something complex nicely you should do so by checking if the logger is enabled for that logging level:
```python
        if log.isEnabledFor(logging.DEBUG):
            log.debug("Registered font families: %s", ", ".join(loaded_fonts))
```
Usually however only done for much heavier formatting.

Sasbom replied:
noted, I pretty much lifted this from usdrecord.
Will be fixed in next push

BigRoy reviewed (commented):
Some first remarks.

Sasbom commented on usd_qtpy/render_util/__init__.py:
```diff
@@ -0,0 +1,2 @@
+# module named render_util to not collide in names with houdini's render.py
```
I'm usually just running __main__.py from the hython environment, that could be where the name conflict happens.
The name wouldn't collide if it was imported as a package so from now on, I'll just run as instructed in the readme.

BigRoy replied:
You can also do `import usd_qtpy.__main__` if you really want to go through main I suspect?
But if you have a need for that it means that I might have just missed exposing something useful elsewhere?

Sasbom replied:
> But if you have a need for that it means that I might have just missed exposing something useful elsewhere?

No, I'm literally running it that way so I don't accidentally push my test run script where i would import it.
I'm not missing any functionality for now but if I do I'll let you know immediatly!

Sasbom commented on usd_qtpy/render_util/playblast.py:
```diff
@@ -0,0 +1,166 @@
+from pxr import UsdAppUtils
+from pxr import Tf, Sdf
+
+from pxr.Usdviewq.stageView import StageView
```
For now, as for this side of things, I only have the stageview in there for getting the scene camera and doing some other things. I could change the import statement to live in function scope of the functions that it lives in, if need be.

The thing I primarily use this for is to get the FreeCamera, which isn't always desirable.

Sasbom commented on usd_qtpy/render_util/playblast.py:
```diff
@@ -0,0 +1,166 @@
+from typing import Union
+import logging
+
+def _setupOGLWidget(width : int, height : int, samples : int = 4):
```
I had a work today, my visual symptoms tend to play up and then I tend to seperate code a lot by adding whitespace. I do this in every language I program in and ... it sucks
I will refactor all function typehints to conform to the standard.
I also renamed the function as suggested, will all be in next commit.

Sasbom commented on usd_qtpy/render_util/playblast.py:
```diff
@@ -0,0 +1,166 @@
+    stageview.ExportFreeCameraToStage(stage,name)
+    return UsdGeom.Camera.Get(stage,Sdf.Path(f"/{name}"))
+
+# Source: UsdAppUtils.colorArgs.py
+def getColorArgs():
+    return ("disabled","sRGB","openColorIO")
```
No, unfortunatly not...

from UsdAppUtils.colorArgs.py:
````py
argsParser.add_argument('--colorCorrectionMode', '-color', action='store',
        type=str, default=defaultValue,
        choices=['disabled', 'sRGB', 'openColorIO'], help=helpText)
```
As you see, it's hard-coded into a list of choices.
If this changes, then we can adapt, but we'd have to make a PR on the usdrecord tool ;P

Sasbom commented on usd_qtpy/render_util/playblast.py:
```diff
@@ -0,0 +1,166 @@
+            logging.error("Recording aborted due to the following failure at time code {0}: {1}".format(timeCode, str(e)))
+            break
+    
+    # Set reference to None so that it can be collected before Qt context.
+    frameRecorder = None
```
I don't quite understand the reasoning behind why they don't `del` the frame_recorder.
I rely on `usdrecord` for guidance, and it specifically made a point to comment about this. 
I will try changing this to see if it has any ramifications down the line, but it'll stay in next cleanup commit.

Sasbom commented on usd_qtpy/style/resources_rc.py:
I have literally no idea, It happened when I synced with the main branch and I then had some errors when pulling.
I'll restore this file in a bit.

Sasbom commented on usd_qtpy/editor.py:
```diff
@@ -94,5 +102,23 @@ def update_panel_checkstate():
+        
+        def render_snapshot(stage,stageview,path):
+            cam = render_util.playblast.cameraFromView(stage,stageview)
+            render_util.playblast.renderPlayblast(stage,path,"1",1920,cam,renderer="GL")
```
I was thinking about this too, about how we would go about managing the hierarchy,
and indeed whether this belongs in the lib.

I feel like if this needs to be added to the lib, that'd make sense, but we should also consider other things to be moved there as well. For example, everything that defines the tree is now in a seperate folder, when one could argue it should be a folder in lib as well, as it serves as utility for the main widgets.

I'd suggest setting up some "rules" on what should and shouldn't be in the library based on functionality,
to me currently it's a bit unclear.

Sasbom commented on usd_qtpy/render_util/playblast.py:
```diff
@@ -0,0 +1,166 @@
+
+    return GLWidget
+
+def findCameras(stage : Usd.Stage, TraverseAll = True) -> list[UsdGeom.Camera]:
+    """
+    Return all camera primitives.
+    TraverseAll is on by default. This means that inactive cameras will also be shown.
+    """
+
+    if TraverseAll:
+        gen = stage.TraverseAll()
+    else: 
+        gen = stage.Traverse()
+    
+    cams = [c for c in gen if UsdGeom.Camera(c)]
+    return cams
```
I changed the functionality to your implementation, I didn't know about UsdGeom.Prim.IsA(sometype)
I made sure to import Generator from `collections.abc` as I know that `typing` will not have it anymore come 3.10 .

Sasbom committed:
house cleaning, suggested bugfixes

Sasbom commented:
Should be good to go. I implemented the majority of the suggestions!
Thank you for grading my homework!

Sasbom committed:
slowly but surely adding methods to deal with cameras

Sasbom committed:
added distance calculation function while downstairs neighbours are drilling in walls. Might need to check this later.

Sasbom committed:
finished clipping planes function.

Sasbom committed:
finished calculate camera position function.

Sasbom committed:
cross platform catch for non-OGL native systems (macosx)

Sasbom committed:
Completed routine to add a framing camera!
You can even set the aspect ratio. Amazing.
There is still some stuff to be done to make a turntable work, but this
is an important step on the road there.

Sasbom committed:
Merge branch 'BigRoy:main' into enhancement/usdview_render_turntable

Sasbom commented:
![framingcam_data](https://github.com/BigRoy/usd-qtpy/assets/87427129/0dfeb0b7-c48c-41b1-9701-15d0f471ac27)
![framingview01 00](https://github.com/BigRoy/usd-qtpy/assets/87427129/25d1ba14-f5a2-49ec-b950-e4b281659eda)
![snapshotframecam](https://github.com/BigRoy/usd-qtpy/assets/87427129/f7c179c3-74ee-4ff8-9dec-4aedd0b04335)

framing cam logic is finally implemented with extensive commenting. I feel like table turning and turn tabling should just be a Xform prim or something on 0,0,0 doing a turn for some amount of frames, so I'll look into generating animation data.

Sasbom committed:
added fit parameter to add some padding

Sasbom commented:
![framingview01 00_fit1 0](https://github.com/BigRoy/usd-qtpy/assets/87427129/4bf921fa-7d8c-4f93-bb64-81d0ddc3ba25)
![framingview01 00_fit1 5](https://github.com/BigRoy/usd-qtpy/assets/87427129/6bb83754-0a0a-4752-9cd1-cb2f417ff449)
Added fit parameter upon request, 
fit = 1 vs fit = 1.5

Sasbom committed:
micro commit to change a variable name in a function to something more sensible.

Sasbom committed:
Created render_util.turntable,
this exposes utility functions for generating things that turn tables
and table turns, not just in cameras, but any prim that is Xformable.

Created render_util.turntable,
this exposes utility functions for generating things that turn tables
and table turns, not just in cameras, but any prim that is Xformable.

Sasbom committed:
bugfixes + exposed turntable render as temp functionality.

Sasbom commented:
![turningtableview_1](https://github.com/BigRoy/usd-qtpy/assets/87427129/e228faca-1136-44aa-abd1-8236ba0a2f7d)
![turnturntableturntbaletableturn](https://github.com/BigRoy/usd-qtpy/assets/87427129/c8b35c3f-90dd-4710-a5f3-81721f049e7b)
![turntablecam_data](https://github.com/BigRoy/usd-qtpy/assets/87427129/8f682632-c229-44da-945e-a86d5b516660)
![cube_parent2prim](https://github.com/BigRoy/usd-qtpy/assets/87427129/8eed256c-f9ac-410c-a473-d9c17bb1f4dc)
turntableturntableturntableturntableturntableturntableturntableturntableturntableturntableturntable

Sasbom committed:
bugfix: adjusted transforms when framing camera as turntable camera

Sasbom commented:
![centered_turningtableview_1](https://github.com/BigRoy/usd-qtpy/assets/87427129/a2342f95-d9cc-49ae-8e61-f5538d958a8a)
turntableturntableturntableturntableturntableturntableturntableturntableturntableturntableturntableturn

now gloriously centered by ensuring that double transformation doesn't take place!

Sasbom committed:
added some utility to deal with the frame argument

Sasbom committed:
made tuples_to_frames_string.tuple_gen way more sane

Sasbom committed:
made functions way less ridiculous

BigRoy commented on usd_qtpy/render_util/playblast.py:
```diff
@@ -0,0 +1,215 @@
+    from pxr.UsdUtils import TimeCodeRange
+    range_token = TimeCodeRange.Tokens.RangeSeparator if end_time is not None else ""        # ":"
+    stride_token = TimeCodeRange.Tokens.StrideSeparator if frame_stride is not None else ""  # "x"
+    
+    end_time, frame_stride = end_time if end_time is not None else "", frame_stride if frame_stride is not None else ""
+
```
This really is more readable as:

```suggestion
    if end_time is None:
        end_time = ""
        
    if frame_stride is None:
        frame_stride = ""
```

BigRoy commented on usd_qtpy/render_util/playblast.py:
```diff
@@ -0,0 +1,215 @@
+    'start_time', 'start_time:end_time', 'start_time:end_timexframe_stride'
+    as defined by the USD standard.
+    """
+    # Keep adhering to USD standard as internally defined.
+    from pxr.UsdUtils import TimeCodeRange
+    range_token = TimeCodeRange.Tokens.RangeSeparator if end_time is not None else ""        # ":"
+    stride_token = TimeCodeRange.Tokens.StrideSeparator if frame_stride is not None else ""  # "x"
+    
+    end_time, frame_stride = end_time if end_time is not None else "", frame_stride if frame_stride is not None else ""
+
+    return f"{start_time}{range_token}{end_time}{stride_token}{frame_stride}"
```
I wonder if this wouldn't be more readable:
```suggestion
    # Keep adhering to USD standard as internally defined.
    from pxr.UsdUtils import TimeCodeRange

    parts = [start_time]
    if end_time is not None:
        parts.append(TimeCodeRange.Tokens.RangeSeparator)  # ":"
        parts.append(end_time)
        
        if frame_stride is not None:
            parts.append(TimeCodeRange.Tokens.StrideSeparator)  # "x"
            parts.append(frame_stride)
            
    return "".join(parts)
```

Also, frame stride would mean nothing if no `end_time` is provided, right? So it should be nested within like this suggestion?

Sasbom replied:
yes, I had it branched like this before but got obsessed trying to reduce the amount of lines.
I do like aliasing the tokens because the namespaces are so ridiculously long.
Next push refactored as:

```py
from pxr.UsdUtils import TimeCodeRange
    range_token = TimeCodeRange.Tokens.RangeSeparator   # ":"
    stride_token = TimeCodeRange.Tokens.StrideSeparator # "x"
    
    collect = f"{start_time}"

    if end_time is not None:
        collect += f"{range_token}{end_time}"
        if frame_stride is not None:
            collect += f"{stride_token}{frame_stride}"
    
    return collect
```

BigRoy replied:
Feel free to "mark resolved" the conversations here which you feel you have fixed in the code already.

Sasbom replied:
> Feel free to "mark resolved" the conversations here which you feel you have fixed in the code already.

I can't unfortunatly do that myself from here! I lack the permissions to do so, it seems.

BigRoy replied:
Thanks - I didn't know. :)

BigRoy commented on usd_qtpy/editor.py:
```diff
@@ -94,5 +102,46 @@ def update_panel_checkstate():
+        def snap_framingcam(stage):
+            from pxr import Sdf
+            framecam = render_util.create_framing_camera_in_stage(stage, Sdf.Path("/"),fit = 1.1)
+            render_util.render_playblast(stage, R"C:\dump\framingview##.##.png", "1", 1920, renderer="GL", camera = framecam)
```
Let's start refactoring out the hardcoded paths to make this more testable already.

You can use something along the lines of

```python

            filename, _selected_filter = QtWidgets.QFileDialog.getSaveFileName(
                parent=self,
                caption="Save frame"
                filter="PNG (*.png); "JPG (*.jpg,*.jpeg)"
            )
```

Sasbom replied:
for the `FrameRecorder` to work the paths need to be extant, so I'll provide a check for the folder in the playblast function, and I'll add a quick filedialog as per suggestion.

BigRoy replied:
> extant

Do you mean existent? Only the _folder_ or the full image path?

Sasbom replied:
> > extant
> 
> Do you mean existent? Only the _folder_ or the full image path?

Yes, existent, Excuse my french! It's been a long day.
Only the folder needs to exist. Next push, playblast will check with
```py
# check existence of directory.
if not os.path.exists(os.path.dirname(outputpath)):
    raise FileNotFoundError("Directory not found, must exist before rendering to it.")
```

BigRoy replied:
If the user basically needs to browse to pick a path the folder will already exist. ;)
But an explicit check is nice in the API functions. In these types of error messages, also good to report what directory was searched for so the error message is informative about what actually didn't exist.

Sasbom replied:
> In these types of error messages, also good to report what directory was searched for so the error message is informative about what actually didn't exist.

Will do!

BigRoy commented on usd_qtpy/render_util/framing_camera.py:
```diff
@@ -0,0 +1,270 @@
+
+    centroid = (bounds_min + bounds_max) / 2
+
+    z_up = (_stage_up(stage) == "Z")
```
Redundant parentheses
```suggestion
    z_up = _stage_up(stage) == "Z"
```

Sasbom replied:
Adjusted. I always get a bit paranoid about bools resolving properly, I'll try to be more courageous in my endeavours

BigRoy commented on usd_qtpy/render_util/framing_camera.py:
```diff
@@ -0,0 +1,270 @@
+    xform_cam = UsdGeom.Xformable(cam_prim)
+    xform_cam.AddRotateXOp(XformOp.PrecisionDouble).Set(90)
+
+def camera_apply_translation(camera: UsdGeom.Camera, translation: Gf.Vec3d):
+    """
+    Apply translation to first found translation operation in
+    """
+    
+    from pxr.UsdGeom import XformOp
+    
+    cam_prim = camera.GetPrim()
+    xform_cam = UsdGeom.Xformable(cam_prim)
+    
+    translate_op = None
+    # check for existing translation operation, if not found, add one to stack.
+    for op in xform_cam.GetOrderedXformOps():
+        op: XformOp
+        if op.GetOpType() == XformOp.TypeTranslate:
+            translate_op = op
+            break
+    else:
+        translate_op = xform_cam.AddTranslateOp(XformOp.PrecisionDouble)
+
+    translate_op.Set(translation)
```
As far as I'm aware this isn't technically camera focused - I'd make this generic targeted to `Usd.Prim` instead since it'd work for any `Xformable` prim.

And then just name it `set_translation(prim, translation)`

Sasbom replied:
You're right. After the cleanup push this'll be on the todo list.

Making this function generic means that it probably should live somewhere else.
A generic function that applies translations/rotations/whatevers to a Prim should live in a different system.

As a first setup, applying different kinds of transformations would be a good thing to get down.
Translation, rotation, the whole transformation attribute could be exposed just like maya does.
(without most of the 4th column, only column 4 row 4 for 1/scale should be exposed, i think,
as scaling for the rest of the axes is as simple as multiplying the corresponding axis in the identity matrix.)

We could dynamically build out a list of widgets to edit operations on the stack based on found operations, as well as create a menu for adding them and, possibly, reordering them.

BigRoy commented on usd_qtpy/render_util/framing_camera.py:
```diff
@@ -0,0 +1,270 @@
+    translate_op = None
+    # check for existing translation operation, if not found, add one to stack.
+    for op in xform_cam.GetOrderedXformOps():
+        op: XformOp
```
Are your type hints failing here? Or why is this needed?

Sasbom replied:
It's solely needed for type hints,
the usd stubs i'm using only deduce that GetOrderedXformOps returns a list,
no info about the class. I can remove this but I would like to have consistent type hints throughout the code.
It admittedly does look a bit silly.

BigRoy commented on usd_qtpy/render_util/framing_camera.py:
```diff
@@ -0,0 +1,270 @@
+    # check for existing translation operation, if not found, add one to stack.
+    for op in xform_cam.GetOrderedXformOps():
+        op: XformOp
+        if op.GetOpType() == XformOp.TypeTranslate:
```
This technically wouldn't detect an existing matrix transformation which in itself could be applying translations as well? Or? Not sure if we care - but if that fails it might be a good note for the docstring.

Sasbom replied:
that's true, i'm not sure if it's nessecary to care. 
I am very explicit about clearing all operations that _might_ exist upon camera creation. 
I feel like doing a broader check in turn_tableize_prim might be a good idea though since the prim getting attributes jammed into would exist already.

What I really have to figure out still, is sublayering and getting the main stage under a new Xform,
like the "graft branches" operation in houdini. (not sure about the maya equivalent)

BigRoy replied:
Have a look at at [`parent_prims`](https://github.com/BigRoy/usd-qtpy/blob/main/usd_qtpy/lib/usd.py#L242-L256). Might suffice for your needs.

Sasbom replied:
> Have a look at at [`parent_prims`](https://github.com/BigRoy/usd-qtpy/blob/main/usd_qtpy/lib/usd.py#L242-L256). Might suffice for your needs.

I'll go look at that when implementing a "preset turntable stage" function.

BigRoy commented on usd_qtpy/render_util/framing_camera.py:
```diff
@@ -0,0 +1,270 @@
+
+def _stage_up(stage: Usd.Stage) -> str:
+    return UsdGeom.GetStageUpAxis(stage)
+
```
Ensure PEP08 - two spaces surrounding top-level functions
```suggestion


def _stage_up(stage: Usd.Stage) -> str:
    return UsdGeom.GetStageUpAxis(stage)


```

Sasbom replied:
implemented in next push
added padding in top level function `_setup_opengl_widget` in playblast submodule as well.

BigRoy commented on usd_qtpy/render_util/framing_camera.py:
```diff
@@ -0,0 +1,270 @@
+    hor_aperture = camera.GetHorizontalApertureAttr().Get()
+    ver_aperture = camera.GetVerticalApertureAttr().Get()
+
+    ver_idx = 2 if z_up else 1
```
Let's be explicit - no need to abbreviate here?
```suggestion
    vertical_axis_index = 2 if z_up else 1
```

Sasbom replied:
Good one, my thinking was to try and keep consistent with the other attributes in the same scope.
I'll adjust this one since it's not all that clear.

BigRoy reviewed (commented):
Just some other notes

Sasbom committed:
refactored some things, added dialog helpers

Sasbom committed:
notes

Sasbom commented:
Shameless TODO list (feel free to suggest additions):
- Look into applying preset scenes for turntables
- Material override playblast? ("Clay" render mode)
- Backplate?

Sasbom committed:
top level function padding

Sasbom committed:
Merge branch 'BigRoy:main' into enhancement/usdview_render_turntable

Sasbom committed:
Proposal: add traverse_prim_descendants to lib.usd

Sasbom committed:
cleaned up test

BigRoy commented on usd_qtpy/lib/usd.py:
```diff
@@ -320,3 +321,28 @@ def remove_spec(spec):
         raise TypeError(f"Unsupported spec type: {spec}")
+
+
+def traverse_prim_descendants(stage: Usd.Stage, path: Union[Sdf.Path,str]) -> list[Usd.Prim]:
+    """
+    DFS Iterate through hierarchy of primitive at path.
+    """
+    if isinstance(path,str):
+        path = Sdf.Path(path)
+
+    prim = stage.GetPrimAtPath(path)
+
+    def recurse_children(prim: Usd.Prim, carry: list[Usd.Prim] = None) -> list[Usd.Prim]:
+        if carry is None:
+            carry = []
+        
+        children: list[Usd.Prim] = prim.GetAllChildren()
+        if children:
+            carry.extend(children)
+
+            for child in children:
+                carry = recurse_children(child,carry)
+        
+        return carry
+    
```
Why do you need this? And how does this differ from `Usd.Stage.Traverse`? :)

BigRoy replied:
Or if you want to start from a particular prim, `Usd.PrimRange(prim)`?

Sasbom replied:
> Why do you need this? And how does this differ from `Usd.Stage.Traverse`? :)
> Or if you want to start from a particular prim, `Usd.PrimRange(prim)`?

I thought I needed it to safely reparent all primitives, but that turns out to not be the case.
Rebasing is all I need!

And yes this is EXACTLY the same as PrimRange down to the depth-first traversal, I just didn't know it existed yet, my bad hahaha. Will take it out again. next commit.

Sasbom committed:
struggling to do reparenting properly

Sasbom commented:
I figured it out. I'm going to have to do some referential trickery.

I'm creating a new scene, and then use references to place everything in the right spot.
Over the Turntable.usda, from Prism:
```
#usda 1.0

over "turntable_reference" (
    prepend references = @X:\VAULT_PROJECTS\COLORBLEED\Kitchen_set\Turntable.usda@
)
{
    over "parent"
    {
        over "subject_reference" (
            prepend references = @X:\VAULT_PROJECTS\COLORBLEED\Kitchen_set\Kitchen_set.usd@
        )
        {
        }
    }
}
```

One thing that i'd like to do, is scale the subject to defined bounds and make it conform to scene up, 
because at the moment, that's not happening.

![image](https://github.com/BigRoy/usd-qtpy/assets/87427129/eeb21b7e-d775-4453-8dc8-ea9c6c853a3a)
As you see, the files are loaded into an external usd file just fine.

![mergedscene_ref_notconform](https://github.com/BigRoy/usd-qtpy/assets/87427129/e9b3cebb-e30b-493d-a964-7e6f1314a7ba)
The small white dot in the middle is the turntable scene, completely overwhelmed by the enormous kitchen.

The plan now is:
- Export current stage to temporary file (Export is different than save, won't override)
- Create stage in memory where logic and transformations take place to fit the referenced geometry to the goal up axis + defined bounds.
- run playblast on stage in memory
- clean up stage export after completion

I might wrap this up in a context wrapper but I don't know. 

I would very much like to start writing a transformations module too, as it would come in very handy to start applying transformations in a neat way.

Sasbom committed:
am no longer confused by reparenting, we're gonna do references.

Sasbom committed:
building out new reference based logic for turntable from file functionality

Sasbom committed:
fleshed out turntable_from_file to working order.

Sasbom commented:
<img width="1086" alt="fit_to_bounding_cube" src="https://github.com/BigRoy/usd-qtpy/assets/87427129/c4bdc7c8-378c-40aa-9f79-a5155c12241f">
Stages that use a reference to a file can now be generated succesfully.

First I'll commit to implementing cleanup.

Sasbom committed:
cleanup integration pass 1

Sasbom committed:
Cleanup finalized

Sasbom committed:
cleared up some comments

Sasbom committed:
removed deprecated functions

Sasbom committed:
updated comments

Sasbom committed:
updated comments

Sasbom committed:
Merge branch 'enhancement/usdview_render_turntable' of https://github.com/Sasbom/usd-qtpy into enhancement/usdview_render_turntable

Sasbom committed:
A preset turntable USD file has been added.
Procedure for turntabling from file has been refined, will automatically
select the included turntable file.
Scaling and centering protocol has been fixed and extensively tested for
file turntable.

BigRoy committed:
Cosmetics / cleanup

BigRoy commented on assets/turntable/turntable_preset.usd:
Can you make this a USDA (usd ascii) file?

Sasbom replied:
> Can you make this a USDA (usd ascii) file?

Will do!

BigRoy committed:
Merge remote-tracking branch 'sasbom/enhancement/usdview_render_turntable' into enhancement/usdview_render_turntable_pas2

# Conflicts:
#	usd_qtpy/render_util/turntable.py

Sasbom committed:
Merge pull request #2 from BigRoy/enhancement/usdview_render_turntable_pas2

Render: Cosmetics / cleanup

Sasbom committed:
Changed template file to .usda, fixed up minor things in turntable.

Sasbom committed:
division per element nessecary, fixed string error in dialog

Sasbom committed:
fix unwanted inclusion

Sasbom committed:
Remove prims made in Render menu

Sasbom committed:
clear bbox before stuffing it full of other things

Sasbom committed:
update gitignore to ignore temp folder

Sasbom committed:
fix ignore of temp folder

Sasbom commented:
![turntable_preset](https://github.com/BigRoy/usd-qtpy/assets/87427129/7b9be482-7c7c-480c-9752-9c0f8ccf6db2)
turntableturntableturnturnturn

Sasbom committed:
automated cleanup at the end of turntable render

Sasbom committed:
small changes, tiny checks for turntable from file functionality.

Sasbom committed:
Set up dialog and editor to dislay it
Add README for setting up turntable

Sasbom committed:
Added a guide to setting up your own turntable.

Sasbom committed:
Merge branch 'enhancement/usdview_render_turntable' of https://github.com/Sasbom/usd-qtpy into enhancement/usdview_render_turntable

Sasbom committed:
base layout style for PlayblastDialog

Sasbom committed:
fleshing out dialog for playblasting

Sasbom committed:
further refined playblast menu

Sasbom commented:
![playblastmenu_test1](https://github.com/BigRoy/usd-qtpy/assets/87427129/27ff7bab-4bbe-43c6-9c32-e0bbcac5aed2)

Sasbom committed:
added destination selection

Sasbom commented:
<img width="299" alt="image" src="https://github.com/BigRoy/usd-qtpy/assets/87427129/f64c7c8a-af58-4bef-bf8a-d08f93aa155d">

TODO: 
- Included purposes
- collecting all relevant info and actually making a playblast
- progress bar?

Sasbom committed:
added pre and post hooks to later hook into with the turntable
interface, as most things will be common to eachother.

Sasbom committed:
updated playblast dialog with area to select purposes

Sasbom committed:
fixed some things, added progress bar and actual playblast functionality to PlayblastDialog

Sasbom commented:
<img width="353" alt="playblast_menu_selectres" src="https://github.com/BigRoy/usd-qtpy/assets/87427129/c5a8e334-7a56-4452-8cb8-63dfb61baa90">

![progress_bar](https://github.com/BigRoy/usd-qtpy/assets/87427129/5101b461-cf90-4fcd-b743-f243d979f841)

<img width="302" alt="playblast_menu_complete" src="https://github.com/BigRoy/usd-qtpy/assets/87427129/6af672f9-8600-492c-bd17-6a13ca340980">

![image](https://github.com/BigRoy/usd-qtpy/assets/87427129/089f5361-1a88-4a38-b851-f8c29b7690a7)

Playblasting works. And it has a progress bar!
The wonders of modern technology.

Sasbom commented:
![kitchen_0002](https://github.com/BigRoy/usd-qtpy/assets/87427129/a50ed8e6-2a51-4ebb-b5dc-38ed6bee9283)

![viewerkitchen_0002](https://github.com/BigRoy/usd-qtpy/assets/87427129/5e2e7851-cef3-4a88-bfe9-dc40afe7e1df)

Confirmed that framing camera and viewer camera works as well.
The viewer camera option is not available if the viewer is opened without it's viewer!

Large scenes like the kitchen tend to lag the main window, so unfortunatly the progressbar won't be perfectly happy with very complicated scenes.

Sasbom committed:
modified PlayblastDialog to be more flexible when inheriting

Sasbom committed:
Complexity combobox fixed and moved to be above render engine, seperated camera choice combobox population into seperate function

Sasbom committed:
added support for camera path argument in turntable_from_file

Sasbom committed:
Added search for cameras in stage from file

Sasbom committed:
Pass along RenderReportable in turntable

Sasbom committed:
started implementing turntable dialog

Sasbom commented:
<img width="302" alt="image" src="https://github.com/BigRoy/usd-qtpy/assets/87427129/9c7b0f78-71d1-4322-8aaf-d091c938d3dd">
<img width="299" alt="image" src="https://github.com/BigRoy/usd-qtpy/assets/87427129/7a98431e-9416-4b21-a7cf-4183a5f0ecab">

After adding in some hooks to the normal playblast dialog, I have started implementing a turntable dialog.
It exposes essentially all turntable functionality to the user, even searching through the file for cameras.
You can see that the frame dialog is different, there is no longer a place to select single frames, nor frame ranges,
there is just starting frame, and length, to make things as easy as possible.

I might add a refresh button next to the camera box, for when the turntable textbox is typed in, or listen in on some changed event. I don't want to do this too often though as reading the cameras requires opening the file.

Sasbom committed:
added functionality for updating camera from file when finished typing.

Sasbom committed:
make sure playblast button on turntable doesn't do anything as it's not implemented yet

Sasbom committed:
logic pass TurntableDialog

Sasbom committed:
adress minor fuckup

Sasbom committed:
exposed fit parameter to dialog for generated framing cameras

Sasbom committed:
aux functionality for rendering turntables, laid out functionality

Sasbom committed:
implemented stage rotation turntable in dialog

Sasbom committed:
implemented camera rotation in dialog

Sasbom committed:
added better cleanup for camera around stage turntable in dialog, fixed pattern matching for filename

Sasbom committed:
implemented turntable from file

Sasbom committed:
cleaned up render menu in editor

Sasbom commented:
![image](https://github.com/BigRoy/usd-qtpy/assets/87427129/d557fe9e-0642-43c0-9822-2214bd84459f)

![image](https://github.com/BigRoy/usd-qtpy/assets/87427129/49e32284-4376-4262-901d-057752c7b1c4)

Implemented all 3 modes of turntabling:
- Rotate camera:
creates a temporary Xform with a camera in it that gets swiveled around the scene
- Rotate subject:
Rotates entire subject in a temporary file that stores a reference to a temporary export of the scene.
- Turntable preset:
It's the preset turntabler.

Sasbom committed:
minor cleanup

Sasbom committed:
small fixes

BigRoy commented on .gitignore:
```diff
@@ -40,6 +40,10 @@ Temporary Items
 ######
 .idea/
 
+# TEMP
+######
+temp/
+
```
Can we remove this - this is better left to your global .gitignore maybe?

Sasbom replied:
I will, 
I'll delete the temp folder when I am done with the files in there.
I needed it to stick around for a bit so that I could take a look at the generated USD files in maya/houdini.

BigRoy commented on usd_qtpy/editor.py:
```diff
@@ -94,5 +102,45 @@ def update_panel_checkstate():
+            stage.RemovePrim(camera.GetPath())
+
+        render_snap_with_framingcam = partial(snap_framingcam, self._stage)
+
+        def show_dialog(this, stage, stageview):
+            dialog = render_util.PlayblastDialog(this, stage, stageview)
+            dialog.show()
+
+        def show_ttable_dialog(this, stage, stageview):
+            dialog = render_util.TurntableDialog(this, stage, stageview)
+            dialog.show()
```
Can we refactor `this` to `parent` to stay closer to Qt terminology.

Sasbom replied:
For sure, i was at a bit of a loss for what to call it but that makes 100% sense.

BigRoy commented on usd_qtpy/render_util/__init__.py:
```diff
@@ -0,0 +1,6 @@
+# module named render_util to not collide in names with houdini's render.py
+from .playblast import *
+from .framing_camera import *
+from .turntable import *
+from .dialog import *
+from .basecls import *
```
Not a fan of star imports myself. We should actually picking what we expose and also add a `__all__` definition, no?

Sasbom replied:
I was planning to do that as a last pass, I agree that `__all__` needs to be defined!

This is still pretty common practice for importing submodules into one module, I'm following what I've found,
and what I've learned.
Generally, it's common practice to import everything, then dictate what you want to expose through `__all__`.

BigRoy replied:
Sounds good to me.

BigRoy commented on usd_qtpy/render_util/dialog.py:
```diff
@@ -0,0 +1,889 @@
+    return filename
+
+
+def prompt_output_path(caption="Save frame"):
+    """Prompt for render filename, ensure frame hashes are present in path"""
+    filename, _ = QtWidgets.QFileDialog.getSaveFileName(
+        caption=caption,
+        filter="PNG (*.png);;JPEG (*.jpg, *.jpeg)"
+    )
+
+    # ensure there's at least two or more hashes, if not, put in 4.
+    REGEX_HASH = r"(#{2,})"
+
+    hash_match = re.search(REGEX_HASH, filename)
+    if not hash_match:
+        filename = _rectify_path_framenumberspec(filename)
+
+    return filename
```
This should return early if no filename picked, otherwise canceling the prompt results in `####`

Also, I'd prefer it if picking e.g. `turntable.0001.png` doesn't result in `turntable.1001####.png` but instead: `turntable.####.png`.

Sasbom replied:
That can be done!
I'll implement this asap.

Sasbom replied:
![image](https://github.com/BigRoy/usd-qtpy/assets/87427129/0b99abd1-ea73-4995-87f4-169d8a092f19)

BigRoy commented on usd_qtpy/render_util/dialog.py:
```diff
@@ -0,0 +1,889 @@
+            self.spinbox_frame_end.setDisabled(True)
+            self.spinbox_frame_stride.setDisabled(True)
+
+    def _prompt_renderoutput(self):
+        filename = prompt_output_path("Render result to...")
+        if filename:
+            self.txt_filename.setText(os.path.normpath(filename))
```
If currently a path is set in the text field it should actually default to that path when browsing again.

Sasbom replied:
Will implement, after checking path validity of course.

BigRoy commented on usd_qtpy/render_util/dialog.py:
```diff
@@ -0,0 +1,889 @@
+    def _gather_purposes(self) -> list[str]:
+        purposes = []
+        
+        if self.chk_purpose_default:
+            purposes.append("default")
+        
+        if self.chk_purpose_render:
+            purposes.append("render")
+
+        if self.chk_purpose_proxy:
+            purposes.append("proxy")
+
+        if self.chk_purpose_guide:
+            purposes.append("guide")
```
As far as I know this doesn't actually check the checkstate?

```suggestion
        if self.chk_purpose_default.isChecked():
            purposes.append("default")
        
        if self.chk_purpose_render.isChecked():
            purposes.append("render")

        if self.chk_purpose_proxy.isChecked():
            purposes.append("proxy")

        if self.chk_purpose_guide.isChecked():
            purposes.append("guide")
```

Sasbom replied:
Very true. I wrote this at a time where I was setting up the interface and I feel like I ended up being a bit blind-
Thanks for spotting this for me.
I'll implement it in next cleanup push.

BigRoy commented on usd_qtpy/render_util/dialog.py:
```diff
@@ -0,0 +1,889 @@
+        """
+        Update progress bar with reported maximum frames
+        """
+        #self._total_frames = frames
```
```suggestion
```

BigRoy commented on usd_qtpy/render_util/dialog.py:
```diff
@@ -0,0 +1,889 @@
+        """
+        Increment progress bar with reported rendered frames.
+        """
+        #self._render_progress = frame
```
```suggestion
```

BigRoy commented on usd_qtpy/render_util/dialog.py:
```diff
@@ -0,0 +1,889 @@
+
+        opt_index = self.cbox_framerange_options.currentIndex()
+
+        if opt_index == 0:
+            return f"{start}"
+        elif opt_index == 1:
+            return playblast.get_frames_string(start,end,stride)  
+        elif opt_index == 2:
+            if self._stageview:
+                return playblast.get_stageview_frame(self._stageview)      
```
This could be clearer if commented or checking against constants?

Sasbom replied:
I don't like setting up constants in classes, especially not for this particular element since this element doesn't
stay the same when inherited, so I'll comment on it. 

Normally I'd define an enum, if something is common.

BigRoy commented on usd_qtpy/render_util/dialog.py:
```diff
@@ -0,0 +1,889 @@
+            # Camera is in scene, doesn't need to be deleted.
+            return data, False
+
+        elif "Framing" in box_text:
+            camera = framing_camera.create_framing_camera_in_stage(
+                self._stage,
+                name="Playblast_framingCam",
+                fit=self.spinbox_fit.value(),
+                width=width,
+                height=height
+                )
+            return camera, True
+        elif "Viewer" in box_text:
```
These should not do a `in` check but just match the full label, no?

Sasbom replied:
I will match the full label, I've done something similar in other sections.
This was written when I wasn't yet sure about what to call the label yet, but I knew that it was going to have the keywords "Framing" and "Viewer" in there.
Will change in next push.

BigRoy commented on usd_qtpy/render_util/dialog.py:
```diff
@@ -0,0 +1,889 @@
+
+        path = self.txt_filename.text()
+
+        print(frames,camera,path)
```
```suggestion
```

BigRoy commented on usd_qtpy/render_util/dialog.py:
```diff
@@ -0,0 +1,889 @@
+        if not path:
+            return
+
+        playblast.render_playblast(self._stage,
+                                   path,
+                                   frames,
+                                   self.spinbox_horresolution.value(),
+                                   camera,
+                                   self.cbox_complexity.currentText(),
+                                   self.cbox_renderer.currentText(),
+                                   "sRGB",
+                                   self._gather_purposes(),
+                                   self
+                                   )
```
This is definitely clearer with the arguments named.

Sasbom replied:
I'll seperate the getting of the arguments from their respective UI elements and store them,
and also name them, `argname=argvariable`

BigRoy commented on usd_qtpy/render_util/dialog.py:
```diff
@@ -0,0 +1,889 @@
+        
+        self.progressbar.setFormat("Rendered %v frames!")
+
+        # cleanup camera if needed
+        if should_destroy:
+            self._stage.RemovePrim(camera.GetPath())
```
Any destroying I'd say is best left to a context manager if needed to _ensure_ the scene is left unaltered.

Sasbom replied:
I have been thinking about that, but there are cases where a camera for example is not supposed to be deleted, because it exists in the stage and is pulled from there. I could make the contextmanager accept an argument that would delete the camera, but that wouldn't make that much of a difference. 
Let me know if you'd still like to see a context manager for camera's, specifically.

What I think could definitely benefit from a context manager, is temporary scene and associated file  operations. I do plenty of that in the turntabling setup, and have been thinking about wrapping those in a context manager.

BigRoy commented on usd_qtpy/render_util/dialog.py:
```diff
@@ -0,0 +1,889 @@
+            # get camera state, to create a valid camera
+            if path:
+                # take new root into account, split off the old scene root.
+                print("not generated cam",path.pathString)
```
```suggestion
```

BigRoy commented on usd_qtpy/render_util/dialog.py:
```diff
@@ -0,0 +1,889 @@
+        vlayout.addLayout(pre_form)
+        self.vlayout.addSpacing(15)
+
+    def playblast_callback(self):
```
This has a lot of logic which looks related to the actual turntabling inside the UI logic. What do you think?

Sasbom replied:
It's mostly to keep the turntable functionality itself as pure as possible. 
The UI is for turntabling, therefore it needs to reflect what can and can be done, from a UI level, however, the functions can still be freestanding. 
Most of the logic is validity checking of user input that happens in the UI, so I feel it is not wrong to keep that logic in the UI itself.

BigRoy commented on usd_qtpy/render_util/dialog.py:
```diff
@@ -0,0 +1,889 @@
+                                          )
+            self.progressbar.setFormat("Rendered %v frames!")
+
+
+        # raise NotImplementedError()
```
```suggestion
```

BigRoy commented on usd_qtpy/render_util/playblast.py:
```diff
@@ -0,0 +1,315 @@
+    cameras = [c.GetPath() for c in iter_stage_cameras(stage)]
+    del stage
+    return cameras
+
```
```suggestion


```
Please check your linter - spacing needs correcting

BigRoy commented on usd_qtpy/render_util/turntable.py:
```diff
@@ -0,0 +1,374 @@
+
+    DOES NOT APPEAR TO WORK, EVEN WITH CORRECT METADATA.
+    """
+    stage = Usd.Stage.CreateInMemory(path)
```
This is not _loading_ the file. It's creating an empty stage in memory that _will_ save to the given path when saved.

As such, it makes sense there's never authored time code range. You might be looking for [`Usd.Stage.Open()`](https://openusd.org/dev/api/class_usd_stage.html#abd74eb9ec70572c2dadaf4cd07144380)?

Sasbom replied:
I thought I already tried that but I could be wrong.

Either way, I've found that even with explicitly setting the layer attribute in e.g. Houdini,
the authored time code range still comes out wrong, however, that was on 19.5, I've yet to check on 20.

BigRoy replied:
Could you provide an example minimal usda file?

Sasbom replied:
I mostly checked with the turntable, but sure, I can provide you with a minimal usda file either this afternoon or tomorrow. I still have HKU to attend to today.

BigRoy commented on usd_qtpy/render_util/turntable.py:
```diff
@@ -0,0 +1,374 @@
+    if index := layer_editor.view.selectedIndexes():
+        layertree_index = index[0]
+    else:
+        layertree_index = layer_editor.view.indexAt(QtCore.QPoint(0, 0))
```
This doesn't feel safe as margins/padding or whatever might make the index not be at the 0, 0 point. I think you might need to go through the model instead getting the first row?

Sasbom replied:
Margins and padding will not affect the point where the index is at,
the is point is in "list space", meaning that it pertains to the model, not the location.

Even so, I'm not sure I want to keep this function in there. 
I thought I needed it, but I haven't used it at all because I stopped using the reparenting function.
instead, if I need a hierarchy to work a certain way I generate elements in order or, if I can't
I'll parent them. This saves me from having to pass in another parameter as well, which is always preferred.

BigRoy replied:
In that, if unused, case let's remove it.

Sasbom replied:
I'll remove it along with some other unused functions.
I had a different approach to rotating the subject at first, but the one I've landed on right now works 100% of the time

BigRoy reviewed (commented):
UIs look good.

Some code remarks added. Let me know what makes sense and whether you need me to refactor things instead.

Sasbom committed:
First pass of cleanup.
TODO: context managers for temporary cameras/scenes

Sasbom committed:
added context managers for managing open scenes and deleting files

Sasbom committed:
ExitStack shenanigans

Sasbom committed:
ExitStack shenanigans part 2: the big bugfix

Sasbom committed:
fixed potential -1 index on camera box in playblast dialog

Sasbom committed:
removed unneeded import to weakref

Sasbom committed:
exposed select functions in `__all__` and added missing stage context manager in playblast.get_file_cameras

Sasbom committed:
fixed spacing from top

Sasbom committed:
removed unnessecary import

Sasbom committed:
modified size, disallowed resizing of playblast/turntable dialogs

Sasbom committed:
bugfix, making sure that right parameters get passed around and that data is collected before files are deleted.

BigRoy commented:
I seemed to be unable to push into this branch - unfortunately. So I've opened a replacement branch with this PR included and some more cosmetic tweaks done on top of it. Closing this in favor of https://github.com/BigRoy/usd-qtpy/pull/43
